### PR TITLE
adding exclude flag to plan and apply to exclude specified resource from map

### PR DESCRIPTION
**Pull request by josephholsten**
Rebase of #2326

### CODE REVIEW

**Comment by phinze** [comment]
Hoping to take this for a spin this week. :+1:

**Comment by josephholsten** [comment]
@phinze rebased this back onto master, let's see what the tests say

**Comment by metcalfc** [comment]
Any update here? This would really help me. I'm using `remote_state` to store the state of a bunch of shared infra (route53, etc) and when I run a `terraform plan -destroy` I'd love to exclude just the state the burn the rest down. If someone nukes the `remote_state` its 😢

**Comment by holybit** [comment]
👍 as I could use this feature for certain. Any progress?

**Comment by carrete** [comment]
> 👍 as I could use this feature for certain. Any progress?

Same.

**Comment by dominik-lekse** [comment]
Hi @phinze, @josephholsten, during developing infrastructure with Terraform this is a feature I currently miss. Besides a pending conflict resolution, is there anything preventing us to merge this functionality, i.e. the general idea/implementation of this feature?

**Comment by josephholsten** [comment]
well, I've rebased this but the tests are still failing. That's the next step.

File: command/plan.go [hunk 1]
```diff
@@ -171,7 +171,8 @@ Options:
 
   -target=resource    Resource to target. Operation will be limited to this
                       resource and its dependencies. This flag can be used
-                      multiple times.
+                      multiple times. Prefixing the resource with ! will
+											exclude the resource.
```
Comment (josephholsten) [inline command/plan.go]: todo: fix

File: terraform/transform_targets.go [hunk 2]
```diff
@@ -70,27 +79,40 @@ func (t *TargetsTransformer) Transform(g *Graph) error {
 			if vr, ok := v.(RemovableIfNotTargeted); ok {
 				removable = vr.RemoveIfNotTargeted()
 			}
-			if removable && !targetedNodes.Include(v) {
-				log.Printf("[DEBUG] Removing %q, filtered by targeting.", dag.VertexName(v))
-				g.Remove(v)
+			if removable {
+				if targetedNodes.Len() > 0 && !targetedNodes.Include(v) {
```
Comment (josephholsten) [inline terraform/transform_targets.go]: `targetedNodes.Len() > 0` breaks `TestCloseProviderTransformer_withTargets`

**Comment by purkhusid** [comment]
Any progress on this?

**Comment by shantanugadgil** [comment]
My use case is that  there are some plan errors in an independent resource, and for the time being I want to just exclude it and see if everything else is fine.

**Comment by rhodrid** [comment]
I could really do with this feature right now

**Comment by protosam** [comment]
Is this still being worked on? My team could use the exclude flag for terraform plans that we're running nightly to monitor drift in our terraform repo.

We have a few items that are always going to be `update in-place`. It would be convenient to just exclude the things we expect to be like that from what's being monitored.

**Comment by zoobab** [comment]
Any update?

**Comment by anupkulkarni24** [comment]
Any Update?

**Comment by dbaumgarten** [comment]
This PR has been open quite a while now. Are there any plans to complete it? It would be a super usefull feature.

**Comment by shantanugadgil** [comment]
4 years!!! Wow! terraform went from separate plugins to single binary and back to providers!☺️☺️☺️

**Comment by describeme** [comment]
bump

**Comment by HeWhoWas** [comment]
This ticket could have started school if it was a particularly gifted child.

**Comment by protosam** [comment]
Depends on the state. In North Carolina, I was 5 when I started school. I'm very surprised how long running this is.

**Comment by sirlatrom** [comment]
I'm going to unsubscribe due to this PR having been converted to a comedian school.

**Comment by protosam** [comment]
@apparentlymart are we still waiting on you to review this? The`license/cla` check failed, who needs to do that agreement? Does @josephholsten need to agree to the contributor license agreement?

This feature would still be nice to have.

**Comment by josephholsten** [comment]
I've just submitted my CLA.

**Comment by Omarimcblack** [comment]
@josephholsten any chance you could fix the conflicts - would be cool to have this implemented https://github.com/hashicorp/terraform/issues/2253

**Comment by josephholsten** [comment]
@Omarimcblack No, as you can see from this ticket's history, I have rebased this many times already. I didn't even write this code in the first place.  @apparentlymart knows how to get ahold of me if hashicorp really wants this.

**Comment by JnMik** [comment]
So @josephholsten , signing that CLA yet ?  :D

**Comment by josephholsten** [comment]
Yes, I have signed previous versions of the CLA and I have signed this version. But I’m not the commit author.

**Comment by jeffery** [comment]
Only 6700 commits behind master. @nevins-b Are you around to help move this forward?

**Comment by tjhiggins** [comment]
@josephholsten could you submit a new pr so you can sign? Our team would find this incredibly useful.

**Comment by josephholsten** [comment]
@tjhiggins 
> Yes, I have signed previous versions of the CLA and I have signed this version. But I’m not the commit author.

**Comment by mildwonkey** [comment]
hi! I am cleaning up some older PRs and am going to close this one. Please do chime in on the linked issue if you are still interested - we haven't forgotten this request, but it's going to take some careful design and isn't on our immediate roadmap. Thanks!

**Pull request closed without merging**